modules/collator-selection/src/lib.rs

```diff
@@ -282,34 +288,22 @@ pub mod pallet {
 		pub fn register_as_candidate(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
 
-			// ensure we are below limit.
-			let length = <Candidates<T>>::decode_len().unwrap_or_default();
-			ensure!(
-				(length as u32) < Self::desired_candidates(),
-				Error::<T>::MaxCandidatesExceeded
-			);
-			ensure!(!Self::invulnerables().contains(&who), Error::<T>::AlreadyInvulnerable);
-
 			let deposit = Self::candidacy_bond();
-			let incoming = CandidateInfo {
-				who: who.clone(),
-				deposit,
-			};
+			let bounded_candidates_len = Self::do_register_as_candidate(&who, deposit)?;
 
-			let mut bounded_candidates = Self::candidates();
+			Self::deposit_event(Event::CandidateAdded(who, deposit));
+			Ok(Some(T::WeightInfo::register_as_candidate(bounded_candidates_len as u32)).into())
+		}
 
-			if bounded_candidates.iter().any(|candidate| candidate.who == who) {
-				Err(Error::<T>::AlreadyCandidate)?;
-			} else {
-				bounded_candidates
-					.try_push(incoming)
-					.map_err(|_| Error::<T>::MaxCandidatesExceeded)?;
-				T::Currency::reserve(&who, deposit)?;
-				<Candidates<T>>::put(&bounded_candidates);
-			}
+		#[pallet::weight(T::WeightInfo::register_candidate(T::MaxCandidates::get()))]
+		pub fn register_candidate(origin: OriginFor<T>, new_candidate: T::AccountId) -> DispatchResultWithPostInfo {
+			T::UpdateOrigin::ensure_origin(origin)?;
 
-			Self::deposit_event(Event::CandidateAdded(who, deposit));
-			Ok(Some(T::WeightInfo::register_as_candidate(bounded_candidates.len() as u32)).into())
+			let deposit: BalanceOf<T> = Default::default();
+			let bounded_candidates_len = Self::do_register_as_candidate(&new_candidate, deposit)?;
```

Comment:
```suggestion
			let bounded_candidates_len = Self::do_register_as_candidate(&new_candidate, Zero::zero())?;
```

Also bit more clear that the deposit is zero